In [4]:
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns

**Data Importing and Pre-processing**

**Import dataset and describe characteristics such as dimensions, data types, file types, and import methods used**

In [27]:
import os
path = os.path.join("..", "data", "raw", "house_sales.csv")
houses = pd.read_csv(path)
houses

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,20141013T000000,221900.0,3.0,1.00,1180.0,5650.0,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000.0,3.0,2.25,2570.0,7242.0,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000.0,2.0,1.00,770.0,10000.0,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000.0,4.0,3.00,1960.0,5000.0,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000.0,3.0,2.00,1680.0,8080.0,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21608,263000018,20140521T000000,360000.0,3.0,2.50,1530.0,1131.0,3.0,0,0,...,8,1530,0,2009,0,98103,47.6993,-122.346,1530,1509
21609,6600060120,20150223T000000,400000.0,4.0,2.50,2310.0,5813.0,2.0,0,0,...,8,2310,0,2014,0,98146,47.5107,-122.362,1830,7200
21610,1523300141,20140623T000000,402101.0,2.0,0.75,1020.0,1350.0,2.0,0,0,...,7,1020,0,2009,0,98144,47.5944,-122.299,1020,2007
21611,291310100,20150116T000000,400000.0,3.0,2.50,1600.0,2388.0,2.0,0,0,...,8,1600,0,2004,0,98027,47.5345,-122.069,1410,1287


In [24]:
print(f"Dataset Dimensions: {houses.shape}")

Dataset Dimensions: (21613, 21)


In [32]:
print("\n--- Column Data Types ---")
print(houses.dtypes)


--- Column Data Types ---
id                 int64
date                 str
price            float64
bedrooms         float64
bathrooms        float64
sqft_living      float64
sqft_lot         float64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object


**Import Method and File Type**: For this project, we are using the King County House Sales dataset. The raw data is stored in a CSV (Comma Separated Values) file. We imported the dataset into our Python environment using the pandas.read_csv() function. To ensure the code is portable across different team members' computers, we utilized the os.path.join() method to define the file path dynamically.

**Dimensions**: The dataset contains 21,613 observations (rows) and 21 features (columns). Each row represents an individual house sale, and the columns represent specific attributes of those houses.

**Data Types**: The features in this dataset are primarily numerical, consisting of 64-bit integers (e.g., id, zipcode, grade) and 64-bit floats (e.g., price, bedrooms, bathrooms).However, we identified that the date column is currently stored as a string (object). As part of our pre-processing, we will need to convert this to a proper datetime format to allow for time-based analysis.


In [36]:
print("--- Missing Values Per Column ---")
print(houses.isnull().sum())

--- Missing Values Per Column ---
id                  0
date                0
price               0
bedrooms         1134
bathrooms        1068
sqft_living      1110
sqft_lot         1044
floors              0
waterfront          0
view                0
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated        0
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64


**Clean, wrangle, and handle missing data**

In [50]:
# 1. Handle Missing Data: Fill with Median

target_columns = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot']
for col in target_columns:
    houses[col] = houses[col].fillna(houses[col].median())

In [45]:
# 2. Wrangle Data: Convert date string to datetime object
houses['date'] = pd.to_datetime(houses['date'])

In [46]:
df = houses

In [51]:
print(df.isnull().sum())

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64


In [52]:
df.head()

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,2014-10-13,221900.0,3.0,1.00,1180.0,5650.0,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,2014-12-09,538000.0,3.0,2.25,2570.0,7242.0,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,2015-02-25,180000.0,2.0,1.00,770.0,10000.0,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,2014-12-09,604000.0,4.0,3.00,1960.0,5000.0,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,2015-02-18,510000.0,3.0,2.00,1680.0,8080.0,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503


**Handling Missing Data**: Our initial scan revealed 1,134 missing values in bedrooms and similar gaps in bathrooms and sqft_living. By applying median imputation, we filled these gaps using a central tendency measure that is resistant to outliers. This ensures that extreme "mansion" prices in the King County dataset do not unfairly influence the replacement values for missing data, maintaining a more realistic distribution for the majority of homes.

**Data Wrangling (Type Conversion)**: Converting the date column from a string to a datetime object is the most significant wrangling step. In its original format, the date was non-calculable text. As a datetime object, we can now extract specific time features (year, month, day) and calculate the Age of the House at Sale, which is a high-impact feature for price prediction.

The post-cleaning verification confirms a 0% null rate across all 21,613 rows. The dataset is now structurally ready for the exploratory data analysis (EDA) phase.